theano/tensor/c_code/alt_blas_template.c

```diff
@@ -156,19 +162,98 @@ void %(name)s(
             PyObject* matrix_C = alt_wrap_fortran_writeable_matrix_%(float_type)s(C, M, N, LDC);
             PyObject* alpha_opA_dot_opB = PyArray_Transpose((PyArrayObject*)opB_trans_dot_opA_trans, NULL);
             if(0 != PyArray_CopyInto((PyArrayObject*)matrix_C, (PyArrayObject*)alpha_opA_dot_opB))
-                alt_fatal_error("NumPy %(name)s implementation: unable to copy ALPHA*op(A)*op(B) into C when BETA == 0.");
+                alt_fatal_error("NumPy %(precision)sgemm_ implementation: unable to copy ALPHA*op(A)*op(B) into C when BETA == 0.");
             Py_XDECREF(alpha_opA_dot_opB);
             Py_XDECREF(matrix_C);
         }
     } else {
         /* C is read, so we must consider it as Fortran-style matrix. */
         PyObject* matrix_C = alt_wrap_fortran_writeable_matrix_%(float_type)s(C, M, N, LDC);
         PyObject* opA_dot_opB = PyArray_Transpose((PyArrayObject*)opB_trans_dot_opA_trans, NULL);
-        alt_numpy_matrix_extended_sum_inplace_%(float_type)s(ALPHA, (PyArrayObject*)opA_dot_opB, BETA, (PyArrayObject*)matrix_C);
+        alt_numpy_matrix_extended_sum_inplace_%(float_type)s(ALPHA, (PyArrayObject*)opA_dot_opB,
+                                                             BETA, (PyArrayObject*)matrix_C);
         Py_XDECREF(opA_dot_opB);
         Py_XDECREF(matrix_C);
     }
     Py_XDECREF(opB_trans_dot_opA_trans);
     Py_XDECREF(opB_transposed);
     Py_XDECREF(opA_transposed);
 }
+
+/* gemv */
+void %(precision)sgemv_(
+    char* TRANS,
+    const int* M,
+    const int* N,
+    const %(float_type)s* ALPHA,
+    %(float_type)s* A,
+    const int* LDA,
+    %(float_type)s* x,
+    const int* incx,
+    const %(float_type)s* BETA,
+    %(float_type)s* y,
+    const int* incy
+) {
+    /**
+    If TRANS is 'n' or 'N', computes:
+        y = ALPHA * A * x + BETA * y
+    Else, computes:
+        y = ALPHA * A.T * x + BETA * y
+    A is a M*N matrix, A.T is A transposed
+    x, y are vectors
+    ALPHA, BETA are scalars
+    **/
+    if (*M < 0 || *N < 0 || *LDA < 0 | *incx < 0 || *incy < 0)
+        alt_fatal_error("The integer arguments passed to %(precision)sgemv_ must all be at least 0.");
+    int transpose = alt_trans_to_bool(TRANS);
+    if (*M == 0 || *N == 0) {
```

Comment:
I don't think this is always True. y size of M. So if M is 0, this is True. But if M is not like and N is zero, y isn,t 0. Remove `|| *N == 0` and update the code bellow and it is good I think.

Did you ran all tests under tensor/tests/test_blas*.py? in a config that don't have blas? Maybe you need to update some tests to don't be skipped.

Reply:
If we don't transpose A, then:
- `A * x` is of size `(M,N) * (N,1) => (M,1)`, and then y is expected to have size M.

But if we transpose A (ie. trans != 'n' and trans != 'N'), then:
- `A * x` is of size `(N,M) * (M,1) => (N,1)` and then y is expected to have size N.

So y should have size M (and `x` with size N) if we don't transpose, but y should have size N (and `x` with size M) if we transpose.

According to this documentation: http://www.netlib.org/lapack/explore-html/d6/d30/group__single__blas__level2_gafc92361b74c6d41c7e5afa0aa5d13ec9.html#gafc92361b74c6d41c7e5afa0aa5d13ec9

Indeed, there are skipped tests in `test_blas*`. I will activate them.

Reply:
All skipped tests check if BLAS C ops are inserted (eg. CGemv), but these ops seems to not be inserted when ldflags is empty, so tests will fail if they are not skipped.

Do we want to insert CGemv even if ldflags is empty (as it has now a fallback implementation) ?

Reply:
yes. @lamblin do you agree with that?

Reply:
In fact, I just got an hesitation. Can you at least do it locally to make sure all the tests pass?
I'm not convinced the c code will be faster then the python code that call blas. The python code have more overhead, but a direct call to blas could be faster

Reply:
We are so close to a release. I prefer to don't enable CGemv in those case by default for now. But can you enable it locally and run the tests to make sure the C code is good?

Reply:
Ok, I will test it locally! Report to coming.

Reply:
With ldflags empty, CGemv doesn't compile because its needs `[sd]dot_`, which does not have fallback C code yet !

Reply:
Can you quickly modify the cgemv code to raise an error in that case so that it compile or quickly implement it in the op? I don't like having new c code with code path not tested. The conv code don't tests all the code path you added.

Reply:
Anyway, this fallback code will be tested with `theano.tensor.nnet.corr` Ops, as they now use [sd]gemv_ (because of unshared convolutions) and they have to compile whatever ldflags contains.

With ldflags empty, this test passes on my computer (Windows 10 64 bits) and should pass with jenkins on windows too:
```
nosetests -xvs theano\tensor\nnet\tests\test_corr.py
```
```
(python2) C:\Users\notoraptor\mila\dev\git\theano>nosetests -xvs theano\tensor\nnet\tests\test_corr.py
Using NumPy C-API based implementation for BLAS functions.
test_fwd (theano.tensor.nnet.tests.test_corr.Grouped_conv_noOptim) ... ok
test_gradinputs (theano.tensor.nnet.tests.test_corr.Grouped_conv_noOptim) ... ok
test_gradweights (theano.tensor.nnet.tests.test_corr.Grouped_conv_noOptim) ... ok
Tests that basic correlations work for odd and even ... ok
Checks dtype upcast for CorrMM methods. ... ok
Tests correlation where filter dilation != (1,1) ... ok
Tests basic correlation in full mode and case where filter ... ok
test_img_kernel_same_shape (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
test_infer_shape_forward (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
test_infer_shape_gradI (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
test_infer_shape_gradW (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
Tests scenario where filter_shape[1] != input_shape[1] ... ok
test_non_contiguous (theano.tensor.nnet.tests.test_corr.TestCorr2D) ... ok
Tests correlation where the {image,filter}_shape is a Constant tensor. ... ok
Tests correlation where subsampling != (1,1) ... ok
Make sure errors are raised when image and kernel are not 4D tensors ... ok
test_fwd (theano.tensor.nnet.tests.test_corr.TestGroupCorr2d) ... ok
test_gradinputs (theano.tensor.nnet.tests.test_corr.TestGroupCorr2d) ... ok
test_gradweights (theano.tensor.nnet.tests.test_corr.TestGroupCorr2d) ... ok
test_graph (theano.tensor.nnet.tests.test_corr.TestGroupCorr2d) ... ok
test_fwd (theano.tensor.nnet.tests.test_corr.TestUnsharedConv) ... ok
test_gradinput (theano.tensor.nnet.tests.test_corr.TestUnsharedConv) ... ok
test_gradweight (theano.tensor.nnet.tests.test_corr.TestUnsharedConv) ... ok
test_fwd (theano.tensor.nnet.tests.test_corr.TestUnsharedCorr2d) ... ok
test_gradinput (theano.tensor.nnet.tests.test_corr.TestUnsharedCorr2d) ... ok
test_gradweight (theano.tensor.nnet.tests.test_corr.TestUnsharedCorr2d) ... ok

----------------------------------------------------------------------
Ran 26 tests in 90.612s

OK
```

Reply:
This tests some of the code path of the new code, not all of them. For example, I'm pretty alpha/beta is always zeros or non-zeros. It will probably only test with contiguous inputs too.

Reply:
Ok, I see.

> Can you quickly modify the cgemv code to raise an error in that case so that it compile or quickly implement it in the op?

I don't hink I can modify cgemv code to raise an error if [sd]dot_ is not available (or maybe by implementing a fallback [sd]dot_ that always generates an error when it is called ?).

I think I should better make another PR to fix MKL linking, and then add a full implementation of [sd]dot_ to this current PR later (maybe tomorrow ?), after we made the beta2. So, we won't merge this current PR until it can be fully tested.